File: VRSampleApp/VRSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/ViewModel/CustomVideoViewModel.m
//
//  CustomVideoViewModel.m
//  VRSampleApp
//
//  Copyright © 2017 Ooyala Inc. All rights reserved.
//

#import "CustomVideoViewModel.h"


@interface CustomVideoViewModel ()

@property (nonatomic) NSString *pCode;
@property (nonatomic) NSString *embedCode;

@end

@implementation CustomVideoViewModel

#pragma mark - Initialziation

- (instancetype)init {
  if (self = [super init]) {
    _pCode = @"";
    _embedCode = @"";
  }
  
  return self;
}

#pragma mark - Public functions

- (void)pCodeDidUpdated:(NSString *)pCode {
  self.pCode = pCode;
  
  [self updateOpenButtonState];
}

- (void)embedCodeDidUpdated:(NSString *)embedCode {
  self.embedCode = embedCode;
  
  [self updateOpenButtonState];
}

- (void)openButtonDidTapped {
  if (![self.embedCode isEqual:@""] && ![self.pCode isEqual:@""]) {
    __weak typeof(self) weakSelf = self;

    [self.delegate dismisCurrentModule:^(BOOL finished) {
      weakSelf.completionBlock(weakSelf.pCode, weakSelf.embedCode);
    }];
  }
}

#pragma mark - Private functions

- (void)updateOpenButtonState {
  [self.delegate openButtonEnabledChanged:![self.embedCode isEqual:@""] && ![self.pCode isEqual:@""]];
}

@end

File: SwiftVRSkinSampleApp/SwiftVRSkinSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/ViewModel/CustomVideoViewModel.swift
//
//  CustomVideoViewModel.swift
//  SwiftVRSkinSampleApp
//
//  Copyright © 2017 Ooyala, Inc. All rights reserved.
//

// MARK: - Protocol
protocol CustomVideoViewModelDelegate: class {
  func openButtonEnabledChanged(enabled: Bool)
  func dismisCurrentModule(completion: ((_ finished: Bool) -> Void)?)
}

class CustomVideoViewModel {
  // MARK: - Public properties
  weak var delegate: CustomVideoViewModelDelegate?
  var completionBlock: ((_ pCode: String, _ embedCode: String) -> Void)?

  // MARK: - Private properties
  private var pCode: String = ""
  private var embedCode: String = ""

  // MARK: - Public functions
  func pCodeDidUpdated(pCode: String?) {
    self.pCode = pCode ?? ""
    updateOpenButtonState()
  }

  func embedCodeDidUpdated(embedCode: String?) {
    self.embedCode = embedCode ?? ""
    updateOpenButtonState()
  }

  func openButtonDidTapped() {
    if embedCode != "" && pCode != "" {
      delegate?.dismisCurrentModule(completion: { (_) in
        self.completionBlock?(self.pCode, self.embedCode)
      })
    }
  }

  // MARK: - Private functions
  private func updateOpenButtonState() {
    delegate?.openButtonEnabledChanged(enabled: embedCode != "" && pCode != "")
  }
}
